File: Lab/DP-swe_4502/lab1/lab1-js/src/models/Driver.ts
import { User } from './User';
import { Location, PaymentMethod } from '../interfaces';
import { RideType, TripStatus } from '../enums';
import { Rider } from './Rider';
import { Trip } from './Trip';

export class Driver extends User {
    constructor(
        id: string,
        name: string,
        preferredPaymentMethod: PaymentMethod,
        public vehicleType: RideType,
        public location: Location,
        public availability: boolean
    ) {
        super(id, name, preferredPaymentMethod);
        this.setRating(4.5)
    }

    acceptRide(trip: Trip): boolean {
        if (this.availability) {
            trip.assignDriver(this);
            return true;
        }
        return false;
    }

    rateRider(rider: Rider, rating: number): void {
        rider.rating = rating;
    }

    updateLocation(newLocation: Location): void {
        this.location = newLocation;
    }

    startTrip(trip: Trip): void {
        trip.status = TripStatus.Assigned;
    }
}

File: Lab/DP-swe_4502/lab1/lab1-java/src/Driver.java
public class Driver extends User {
    private String vehicleType;
    private String location;
    private boolean isAvailable = true;

    public Driver(String id, String name) {
        super(id, name);
    }

    public String getVehicleType() {
        return vehicleType;
    }

    public void setVehicleType(String vehicleType) {
        this.vehicleType = vehicleType;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void setAvailable(boolean available) {
        isAvailable = available;
    }

    public void acceptRide(Trip trip) {
        if (isAvailable) {
            trip.assignDriver(this);
            System.out.println("Driver " + getName() + " accepted the trip request.");
            setAvailable(false);
        }

    }

    public void cancelRide(Trip trip) {
        trip.setStatus(TripStatus.CANCELLED);
        System.out.println("Driver " + getName() + " cancelled the trip.");
        setAvailable(true);
    }

    public void rateRider(Rider rider, double rating) {
        rider.setRating(rating);
    }

    public void updateLocation(String newLocation) {
        this.location = newLocation;
    }

    @Override
    public void receiveNotification(String message) {
        System.out.println("Notification for driver " + getName() + ": " + message);
    }
}